### DISCUSSION TITLE
UI Framework

### DISCUSSION THREAD

**Discussion by zicklag** [muni-town/weird#19] [General]
For the UI framework in Weird we want to go as simple, portable, and lightweight as possible. We also want try to make decisions that will make it easier to create UI _plugins_ that may be compiled or otherwise developed separately from other UI components.

To that end, we were thinking of using [Web Components](https://developer.mozilla.org/en-US/docs/Web/API/Web_components) ( note this is different than [WASM Components](https://component-model.bytecodealliance.org/language-support/rust.html), which we are also considering, but for the "business logic" ).

[Lit](https://lit.dev/) appears to be the most popular framework of web components, so that is our initial consideration.

## Server-side Rendering

While I didn't initially realize it,server-side-rendering is also important. For example, if we have a list of links for the user's links page, we need that list to be rendered with all it's elements on the server, before it is returned to the client, for proper SEO.

But we also need client-side interactivity. We want to be able to drag-and-drop, and live edit the web page, which means we need some form of JS interactivity, too.

### Server Templates + Dynamic Components

Server-side rendering might be as simple as handlebars-style templates, like Terra which is used for rendering Zola sites. This would be easy to do, but it means that there is a disconnect between the server-side rendering and the client side-rendering. This disconnect can be made to work, but it is not necessarily optimal.

For example, to create the page initially, it will read the list of links from the database, and then use the Terra template to create the initial page HTML. This page can contain Lit web components, but it already has all the links displayed, as static HTML.

Usually, if you wanted to add an item to the list in Lit, you would add some JS data to a list, and have it dynamically add the new HTML element, but since the HTML elements are already statically created, you might have to mess with things to make it dynamic.

## SSR Compatible Web Component Framework

The alternative, is to use a UI framework that we can properly render on the server in WASM, and then update to interactive-mode by client-side "hydration". There are a couple of possibilities that I know of right now.

### Enhance

[Enhance](https://enhance.dev/) is definitely worth looking into. After a quick recon, it looks quite sensible, and very much along the lines of a _simple_ web framework. I like that you can use HTML only, and add JavaScript only if you want to, and, importantly, they support SSR inside of WASM.

If there are no technical barriers it appears to be well aligned with our goals. I think it hits well on the point of, not needing to learn much in addition to normal web stuff.

> **Side Note:** The parent company ( or something ) of Enhance also has a cloud service that might be worth looking into, but I'm not sure that it's what we want. I'm getting a "too enterprise-y and AWS-y" vibe, but that's highly subjective and weakly founded. It could still be totally fine.

### Leptos ( Or Another Rust SSR-Compatible Framework )

[Leptos](https://leptos.dev/) is an up-and-coming Rust framework that [supports SSR](https://book.leptos.dev/ssr/23_ssr_modes.html?highlight=ssr#using-ssr-modes). I don't see any reason yet that Leptos wouldn't be able to do our SSR in WASM, and update to client-side interactive like we need it to.

## Summary

In conclusion, I think we should probably look into one of the SSR compatible web frameworks. I'm leaning towards trying Enhance at this point because it's so simple and so focused on web standards, instead of fancy web frameworks.

Leptos is a compelling option due to being also written in Rust, but there are a handful of reasons I lean away from Leptos:

- Not introducing unnecessary complexity: the less framework we depend on, as long as we can still develop quickly, the less barrier we put in the way of potential contributors, and the less we have to keep up with from a frameworks, compilers, tooling, etc. The less we depend on, the easier it is to change. Since web components are built into browsers now, we get a lot of the values of frameworks, while still being pretty Vanilla.
- Leptos is relatively new: while Leptos and Rust for UI might add value, it's also newer and going to be going through more iterations which we'll have to keep up with, which will take time and mental bandwidth at some point.
- We can always change later, but it's easier to upgrade than to downgrade: if we start off with less, we can always add more, but if we start off with too much, it's not easy to trim down.
- Rust WASM relies on more binding generators and stuff which can cause trouble or just take more work in edge cases when you just want to make something work, especially if we want to integrate with existing JS tooling like [Editable.website](https://editable.website) ( or libraries `editable.website` uses to make the editing work, anyway ).

To clarify, I know that we're already planning on using Rust, with WASM, and binding generators, at least in [service worker](https://github.com/commune-os/weird/discussions/16#service-workers), but when you get into the client-side browser stuff, sometimes it's easier just to be able to use JS for browser stuff.

I also realize that Enhance _is_ another framework. It's not like we're completely getting away from a framework. But I like that it tries to be a _thin_ framework.

I think that doing some quick experimentation with Enhance to see if there are any showstopper limitations would be worth it.

**Comment by erlend-sh** [comment]
What about [Lit SSR](https://lit.dev/docs/ssr/overview/)? Apparently the Eleventy plugin it can integrate with is also quite robust.

**Reply by zicklag** [reply]
Unfortunately LitSSR only works with NodeJS, so we can't run it on our Rust/WASM server.

**Reply by zicklag** [reply]
Enhance also [uses web components](https://enhance.dev/docs/conventions/components#%40enhance%2Fcustom-element), so components still end up portable. It seems that they have also put some thought into things with their component library, trying to avoid some gotchas that pop up with a common kind of Shadow DOM based web components: [Head Towards the Light DOM](https://begin.com/blog/posts/2023-11-10-head-toward-the-light-dom).

I don't have a lot of experience with web components yet, but I'm glad to see that they are thinking about things, trying to actually solve problems, and not just going with what's commonly done.

Not that Lit isn't good, but at least I'm not seeing any big disadvantage to Enhance yet, if we can't use lit.

**Comment by zicklag** [comment]
### Leptos Might be our Best Option

_This was hashed on on Discord [here](https://discord.com/channels/938334993191686174/944525665854701588/1229101712959344651)._

After looking deeper into Enhance, it seems that the only piece we can run on our server in WASM is the element/component rendering piece. This means that all of the conventions for fetching data from the database has to be our own.

That's not necessarily a problem, but I'm wondering at this point whether or not Leptos actually does build a better experience for us, if it means that we can do automatic offline first. The idea is that we will be writing our server-side logic, in Rust, and then automagically running that server-side logic locally to enable offline-first + optimistic updates. So, if we are writing Rust anyway, and leptos does SSR, it might just end up a cleaner solution to use Leptos. That way we're not having to combine Rust and JS conventions...

There is the possiblity of something like using something like [WasmEdge SSR](https://wasmedge.org/docs/develop/javascript/ssr/) to do SSR for a JS framework like Lit. WasmEdge has an example for running JS SSR code for things like React, but at this point it does feel like Rust SSR is probably futher along, and maybe a better fit than JS in WASM.

**Reply by erlend-sh** [reply]
Alright, *all aboard the Pure Rust train,* ***choo choo!*** 🚂

Seriously considering Lit et.al. has been worthwhile though. Once the repo is a bit further along we can add a section for ‘Why not just a simple JS app?’ and explain what we’ve tried so far and what might indeed be possible with JS-in-WASM in the future.
